server/app/views/components/buttons/ButtonAction.java

```diff
@@ -0,0 +1,36 @@
+package views.components.buttons;
+
+import com.google.auto.value.AutoOneOf;
+
+/**
+ * Represents the action that should be taken when the end user clicks a button. As an AutoOneOf,
+ * exactly one option from the enum will be selected at a time.
+ */
+@AutoOneOf(ButtonAction.Action.class)
+public abstract class ButtonAction {
+  public enum Action {
+    REDIRECT,
+    SUBMIT_WITH_FORM_ID,
+    SUBMIT_WITHOUT_FORM_ID
+  }
+
+  public abstract Action action();
+
+  public static ButtonAction ofRedirect(String redirectUrl) {
+    return AutoOneOf_ButtonAction.redirect(redirectUrl);
+  }
+
+  public static ButtonAction ofSubmitWithFormId(String formId) {
```

Comment:
Could you instead overload a method name for the versions with and without form ID, e.g. ofSubmit() and ofSubmit(String formId)?

Reply:
The method names need to match the enum above. But I can change submitWithoutFormId to just submit. Done.